File: dotnet/ThirtyFour.UI.Tests/Window/BrowserWindowTests.cs
using System;
using System.Collections.ObjectModel;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using ThirtyFour.UI.Window;

namespace ThirtyFour.UI.Tests.Window
{
    [TestClass]
    public class WindowTests : BaseTestSuite
    {
        private void QuitExtraDrivers(String currentHandle)
        {
            ReadOnlyCollection<String> handles = driver.WindowHandles;

            foreach (String handle in handles)
            {
                if (!handle.Equals(currentHandle))
                {
                    driver.SwitchTo().Window(handle);
                    driver.Quit();
                }
            }
        }

        [TestMethod]
        public void DriverEqualsCurrentDriver()
        {
            string currentHandle = driver.CurrentWindowHandle;

            driver.FindElement(By.LinkText("Popup")).Click();
            BrowserWindow window = new BrowserWindow(driver);
            Assert.AreEqual(driver, window.Driver);

            QuitExtraDrivers(currentHandle);
        }

        [TestMethod]
        public void ExceptionThrownWhenNoPopupExists()
        {
            string currentHandle = driver.CurrentWindowHandle;

            try
            {
                _ = new BrowserWindow(driver);
                Assert.Fail("Expected exception not thrown");
            }
#pragma warning disable RECS0022 // A catch clause that catches System.Exception and has an empty body
            catch
#pragma warning restore RECS0022 // A catch clause that catches System.Exception and has an empty body
            {
               
            }
            finally {
                QuitExtraDrivers(currentHandle);
            }

        }

        [TestMethod]
        public void DefaultHandleNotEqualsDriverHandleWhenPopupDoesExist()
        {
            string currentHandle = driver.CurrentWindowHandle;

            driver.FindElement(By.LinkText("Popup")).Click();
            BrowserWindow window = new BrowserWindow(driver);
            Assert.AreNotEqual(driver.CurrentWindowHandle, window.DefaultHandle);

            QuitExtraDrivers(currentHandle);
        }

        [TestMethod]
        public void WindowSwitchesToCorrectWindowWhenNoMatcherIsProvided()
        {
            string currentHandle = driver.CurrentWindowHandle;

            driver.FindElement(By.LinkText("Popup")).Click();
            BrowserWindow window = new BrowserWindow(driver);
            new WebDriverWait(driver, TimeSpan.FromSeconds(2)).Until(SeleniumExtras.WaitHelpers.ExpectedConditions.UrlToBe("https://github.com/danhumphrey/ThirtyFour.UI"));

            QuitExtraDrivers(currentHandle);
        }
    }
}

File: java/src/test/java/com/github/danhumphrey/thirtyfour/ui/test/window/BrowserWindowTests.java
package com.github.danhumphrey.thirtyfour.ui.test.window;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import com.github.danhumphrey.thirtyfour.ui.test.BaseTestSuite;
import com.github.danhumphrey.thirtyfour.ui.window.BrowserWindow;

public class BrowserWindowTests extends BaseTestSuite {
	
	@Test
	public void testExceptionThrownWhenNoPopupExists() throws Exception {
		
		WebDriver driver = getDriver();
		String currentHandle = driver.getWindowHandle();
		
		try {
			new BrowserWindow(driver);
			Assertions.fail("Expected exception not thrown");
		} catch(Exception ex) {
		} finally {
			quitExtraDrivers(currentHandle);
		}
	}
	
	@Test
	public void testDriverEqualsCorrectDriver() throws Exception {
		
		WebDriver driver = getDriver();
		String currentHandle = driver.getWindowHandle();
		
		driver.findElement(By.linkText("Popup")).click();
		try {
			BrowserWindow window = new BrowserWindow(driver);
	        Assertions.assertEquals(driver, window.getDriver());
		} finally {
			quitExtraDrivers(currentHandle);
		}
	}
	
	@Test
    public void testDefaultHandleNotEqualsDriverHandleWhenPopupExists() throws Exception {
		WebDriver driver = getDriver();
		String currentHandle = driver.getWindowHandle();
		
        driver.findElement(By.linkText("Popup")).click();
        
        try {
        	BrowserWindow window = new BrowserWindow(driver);
            Assertions.assertNotEquals(driver.getWindowHandle(), window.getDefaultHandle());
            
        } finally {
        	quitExtraDrivers(currentHandle);
        }
        
    }
	
	@Test
	public void testWindowSwitchesToCorrectWindowWhenNoMatcherIsProvided() throws Exception {
		WebDriver driver = getDriver();
		String currentHandle = driver.getWindowHandle();
		
		driver.findElement(By.linkText("Popup")).click();
		
		try {
			new BrowserWindow(driver);
	        new WebDriverWait(driver, 2).until(ExpectedConditions.urlToBe("https://github.com/danhumphrey/ThirtyFour.UI"));
			
		} finally {
			quitExtraDrivers(currentHandle);
		}
    }
}
